src/dsl.jl

```diff
@@ -516,155 +451,216 @@ end
-        metadata = $(rxstruct.metadata)))
-    for sub in rxstruct.substrates
-        push!(reaction_func.args[3].args, sub.reactant)
-        push!(reaction_func.args[5].args, sub.stoichiometry)
+    rx_constructor = :(Reaction($rate, $subs_init, $prod_init, $subs_stoich_init,
+        $prod_stoich_init; metadata = $(rx.metadata)))
+
+    # Loops through all products and substrates, and adds them (and their stoichiometries)
+    # to the `Reaction` expression.
+    for sub in rx.substrates
+        push!(rx_constructor.args[4].args, sub.reactant)
+        push!(rx_constructor.args[6].args, sub.stoichiometry)
     end
-    for prod in rxstruct.products
-        push!(reaction_func.args[4].args, prod.reactant)
-        push!(reaction_func.args[6].args, prod.stoichiometry)
+    for prod in rx.products
+        push!(rx_constructor.args[5].args, prod.reactant)
+        push!(rx_constructor.args[7].args, prod.stoichiometry)
     end
-    reaction_func
+    return rx_constructor
 end
 
-# takes a ModelingToolkit declaration macro like @parameters and returns an expression
-# that calls the macro and saves it in a variable given by namesym based on name.
-# scalarizes if desired
-function assign_expr_to_var(nonempty, ex, name; scalarize = false)
-    namesym = gensym(name)
-    if nonempty
-        if scalarize
-            symvec = gensym()
-            ex = quote
-                $symvec = $ex
-                $namesym = reduce(vcat, Symbolics.scalarize($symvec))
-            end
-        else
-            ex = quote
-                $namesym = $ex
-            end
-        end
-    else
-        ex = :($namesym = Num[])
-    end
-    ex, namesym
+# Recursively escape functions within equations of an equation written using user-defined functions.
+# Does not escape special function calls like "hill(...)" and differential operators. Does
+# also not escape stuff corresponding to e.g. species or parameters (required for good error
+# for when e.g. a species is used as a differential, or for time delays in the future).
+function escape_equation!(eqexpr::Expr, all_syms)
+    eqexpr.args[2] = recursive_escape_functions!(eqexpr.args[2], all_syms)
+    eqexpr.args[3] = recursive_escape_functions!(eqexpr.args[3], all_syms)
+    eqexpr
 end
 
 ### DSL Option Handling ###
 
-# Checks if the `@default_noise_scaling` option is used. If so, read its input and adds it as a
-# default metadata value to the `default_reaction_metadata` vector.
-function check_default_noise_scaling!(default_reaction_metadata, options)
-    if haskey(options, :default_noise_scaling)
-        if (length(options[:default_noise_scaling].args) != 3) # Because of how expressions are, 3 is used.
-            error("@default_noise_scaling should only have a single input, this appears not to be the case: \"$(options[:default_noise_scaling])\"")
-        end
-        push!(default_reaction_metadata.args,
-            :(:noise_scaling => $(options[:default_noise_scaling].args[3])))
```

Comment:
`check_default_noise_scaling!` moved later to reflect the order with which these are called.